Step3 - 질문 삭제하기 리팩터링

wrallee opened the pull request:
안녕하세요
3단계 미션 제출합니다.

히스토리-질문/답변 간 관계에 대해 고민을 많이했는데 제대로 한건지 모르겠군요.
그리고 질문-답변 간 연관관계를 설정하는게 이상한것같기도 해서 계속 바꿨다가 복구했다가 한 것 같습니다.

수업에서 연관관계 편의 메서드는 연관관계의 주인(외래키의 관리자) 쪽에서 관리해주는것이 속편하다고 들어서 답변에서 질문을 세팅하도록 구현했었는데요. 이 부분은 어떻게 생각하시는지 궁금합니다. Question.addAnswer에서 세팅하는게 자연스러울까요?

리뷰 부탁드리겠습니다~!

wrallee committed:
docs(delete): 질문 삭제하기 리팩터링 요구사항 정의

wrallee committed:
test(qna): 질문/답변 단위 테스트 작성

wrallee committed:
feat(qna): 질문/답변 삭제기능 및 히스토리 반환 기능 구현

wrallee committed:
refactor(qnaService): 서비스 메서드 리팩터링

wrallee committed:
refactor(answers): 답변 목록을 일급 컬렉션으로 변경

wrallee committed:
refactor(deleteHistories): 삭제 이력 목록을 일급 컬렉션으로 변경

wrallee committed:
docs(qna): 3단계 기능 구현 완료

parkeeseul commented on src/main/java/qna/service/QnaService.java:
```diff
@@ -34,25 +30,7 @@ public Question findQuestionById(Long id) {
-        deleteHistoryService.saveAll(deleteHistories);
+        deleteHistoryService.saveAll(
+            findQuestionById(questionId).delete(loginUser));
     }
```
한줄로 작성하기 보다는 다음과 같이 적절하게 변수로 분리해주는 것이 좀 더 가독성이 좋을 것 같은데 어떻게 생각하시나요? 😄 
``` java
    @Transactional
    public void deleteQuestion(User loginUser, Long questionId) throws CannotDeleteException {
        Question question = findQuestionById(questionId);
        deleteHistoryService.saveAll(question.delete(loginUser));
    }
```

parkeeseul commented on src/main/java/qna/domain/Answer.java:
```diff
@@ -56,22 +57,38 @@ public Answer(Long id, User writer, Question question, String contents) {
+
+    private void validateQuestion(Question question) {
+        if (Objects.isNull(question)) {
+            throw new IllegalArgumentException("null에 답변을 달 수 없습니다.");
```
`null에 답변을 달 수 없습니다` 라는 내용보다는
`존재하지 않는 질문 입니다` 라는 메세지가 좀더 의미전달을 잘 할 수 있을 것 같은데 어떻게 생각하시나요? 😄

parkeeseul commented on src/main/java/qna/domain/Answers.java:
```diff
@@ -0,0 +1,38 @@
+
+    private void validateAnswer(Answer answer) {
+        if (Objects.isNull(answer)) {
+            throw new IllegalArgumentException("null은 답변으로 추가할 수 없습니다.");
```
해당 에러메세지도 함께 확인 부탁드려요! 😄

parkeeseul commented on src/main/java/qna/domain/Question.java:
```diff
@@ -93,14 +96,37 @@ public boolean isDeleted() {
 
-    public void setDeleted(boolean deleted) {
-        this.deleted = deleted;
+    public void addAnswer(Answer answer) {
+        validateAnswer(answer);
+
+        this.answers.add(answer);
     }
```
질문 주신 사항에 대한 저의 의견 여기에 남겨두겠습니다 ! 😄 

> Q. 수업에서 연관관계 편의 메서드는 연관관계의 주인(외래키의 관리자) 쪽에서 관리해주는것이 속편하다고 들어서 답변에서 질문을 세팅하도록 구현했었는데요. 이 부분은 어떻게 생각하시는지 궁금합니다. Question.addAnswer에서 세팅하는게 자연스러울까요?

A. 설계에 따라 조금 달라질 수 있을 것 같아요! 
연관관계 편의 메서드가 어느 쪽에 위치 하던지 무방할 경우에는 연관관계의 주인 쪽에서 관리해주는 것이 좋겠지만
개인적으로는 지금 처럼 Question이 Answer 포함하는 관계라면 구현해주신 것 처럼 Question에 연관관계 편의 메서드를 위치시키는 것이 객체의 관계에 맞춘 더 자연스러운 설계라고 생각합니다. 😄 

하지만 사람마다 중요하게 생각하는 부분이 다르고 설계나 상황에 따라 더욱 좋은 설계가 있기 때문에 우찬님이 더 좋은 방법이라 생각하시는 대로 작성해보시고 어떠한 차이가 있는지 고민해보시면 좋을 것 같아요!

wrallee replied:
리뷰 감사합니다!
앞선 피드백들 반영하고 좀 더 고민해봐야겠네요
얼른 수정하고 다시 요청드리겠습니다~ 😊

parkeeseul reviewed (changes requested):
미션 잘 진행해주셨네요! 👍 
몇 가지 코멘트와 질문 주신 사항에 대한 저의 의견 남겨두었으니 확인 부탁드립니다! 😄 
JPA  마지막 단계네요! 조금만 더 힘내주세요! 미션 화이팅 입니다! 🔥

wrallee committed:
refactor(qnaService): 질문조회-히스토리저장 라인 분리

한줄로 작성 시 동작 단계에 대한 이해가 어려워질 수 있음

wrallee committed:
refactor(answer): Answer 등록 실패 메시지 수정

변수의 값을 서술하기보다는 설명하도록 수정
값이 null인건 debugger를 통해 확인 가능

wrallee committed:
refactor(qna): 연관관계 편의 메서드 변경

답변과 질문의 연관관계를 설정하는 주체 변경
AS-IS: 답변 생성 즉시 질문과 연관관계 설정
TO-BE: 답변 생성 후 질문에 추가 시 연관관계 설정

wrallee committed:
refactor(answer): 연관 관계 밸리데이션 변경

parkeeseul reviewed (approved):
피드백 잘 반영해주셨네요! 👍  이만 머지 하도록 하겠습니다! 
그동안 많은 피드백 반영해주시느라 정말 고생많으셨습니다! 😄 
그리고 잘 반영해주셔서 감사합니다 🙇‍♀️ 
남은 미션들도 꼭 완주 하셔서 많은 인사이트 얻어가시길 바라겠습니다!
고생 많으셨습니다! 마지막까지 화이팅!!! 🔥
